_posts/2022-09-07-quick-tips-to-speed-up-ci.markdown

```diff
@@ -0,0 +1,145 @@
+
+```bash
+$ yarn jest --shard 1/5
+```
+
+What this says is: take the total number of tests (100), divide them into five
+buckets (containing 20 tests each), and execute the test runner against the
+first bucket (the first 20 tests). Continuing:
+
+```bash
+$ yarn jest --shard 2/5
+```
+
+Now take the second bucket and execute the next 20 tests -- and so on. Simple
+enough.
+
+Taking this further, we could turn this into a bash loop, include an `&` symbol
+to run things in parallel and automate some of the redundancy away:
+
+```bash
+BUCKETS=5
+
+for i in {1..${BUCKETS}}
+do
+   yarn jest --shard $i/$BUCKETS &
+done
+```
+
+For many the above snippet should be sufficient to speed up your test suite, but
+who wants to write bash loops? Thankfully, most modern CI task runners contain
+the ability to split jobs into separate processes programatically and so this
+kind of logic is unnecessary.
+
+Here's how to do this in Circle CI:
+
+```yaml
+test:
+  parallelism: 5
+  steps:
+    - run: yarn test --shard=$(expr $CIRCLE_NODE_INDEX + 1)/$CIRCLE_NODE_TOTAL
+```
+
+Set a `parallelism` value, and drop the `jest` command into a cool one-liner.
+The variable `CIRCLE_NODE_INDEX` refers to which container index the job is
+running on, and `CIRCLE_NODE_TOTAL` points to the value of `parallelism`.
+
+On Artsy.net, we've been able to reduce the average time it takes to run our
+unit tests from around ~10 minutes per PR to just above 2m. A 4-5x performance
+improvement.
+
+## Parallelizing Cypress.io Integration Tests (For Free)
+
+For those who want robust integration test coverage, [Cypress.io][cypress] has
+been a game-changer due to its reliability and ease of use. Here at Artsy we use
+it in a number of apps, most notably [Integrity][integrity]. One complaint,
+however, is just how _slow_ it is. This is reasonable; Cypress is simulating a
+user browsing your website and sometimes a user needs to do `x` and `y` (such as
+logging in) before they can do `z`. At scale this can really slow things down
+and lead to bottlenecks, especially if deploys are dependent on all of your
+integration tests passing.
+
+The Cypress.io team has recognized this bottleneck and released the [Cypress
+Dashboard][dashboard], a paid product which includes the ability to unlock
+parallelized tests on your CI. For those willing to pay for another SAAS product
+this will get the job done well, but for those with leaner budgets there's
+another way to accomplish this for free, and on CircleCI it's very easy to setup
+via the CircleCI CLI command `split`.
+
+You can check out the [full example here][smoketests], but in short:
+
+```yaml
+integration:
+  parallelism: 5
+  run: |
+    TESTS=$(circleci tests glob "cypress/integration" | circleci tests split | paste -sd ',')
+    cypress run --spec $TESTS
+```
+
+We use the `circleci tests glob` command to gather all of our tests, and then
+pipe that into the `split` command which will divide our tests into buckets,
```

Comment:
**suggestion (non-blocking):** Maybe another link here out to [Circle's cli-specific `split` subcommand](https://circleci.com/docs/parallelism-faster-jobs#using-the-circleci-cli-to-split-tests)? I def did a double-take, thinking at first this referred to the classic unix `split` command.